arangod/Pregel/Algos/AIR/AbstractAccumulator.h

```diff
@@ -0,0 +1,193 @@
+////////////////////////////////////////////////////////////////////////////////
+///
+/// DISCLAIMER
+///
+/// Copyright 2020 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Heiko Kernbach
+/// @author Lars Maier
+/// @author Markus Pfeiffer
+///
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef ARANGODB_PREGEL_ALGOS_VERTEX_ACCUMULATORS_ABSTRACT_ACCUMULATOR_H
+#define ARANGODB_PREGEL_ALGOS_VERTEX_ACCUMULATORS_ABSTRACT_ACCUMULATOR_H 1
+#include <Pregel/Algos/AIR/Greenspun/EvalResult.h>
+#include <iostream>
+#include <numeric>
+
+#include "velocypack/Builder.h"
+#include "velocypack/velocypack-aliases.h"
+
+#include "AccumulatorOptionsDeserializer.h"
+#include "MessageData.h"
+
+namespace arangodb {
+namespace pregel {
+namespace algos {
+namespace accumulators {
+class VertexData;
+
+template <typename T>
+class Accumulator;
+
+struct AccumulatorBase {
+  AccumulatorBase(){};
+  virtual ~AccumulatorBase() = default;
+  template <typename T>
+  Accumulator<T>* castAccumulatorType() {
+    return dynamic_cast<Accumulator<T>*>(this);
```

Comment:
Is this cast guaranteed to succeed? If so, should we add a check/assertion here that the cast result is not null?

Reply:
Unused, gone.